.circleci/config.yml

```
@@ -0,0 +1,39 @@
+version: 2.1
+
+jobs:
+  build-test-deploy:
+    machine: true
+    environment:
+      APP_NAME: aws
+      APP_VERSION: 1.16.26
+    steps:
+      - checkout
+      - run:
+          name: Docker version
+          command: |
+            docker info
+            docker --version
+      - restore_cache:
+          key: docker-cache-{{ .Environment.CIRCLE_SHA1 }}
+      - run:
+          name: Build image
+          command: |
+            docker build -t cgswong/${APP_NAME}:${APP_VERSION} .
+      - run:
+          name: Test version
+          command: |
+            docker run cgswong/${APP_NAME}:${APP_VERSION} aws --version
+      - save_cache:
+          key: docker-cache-{{ .Environment.CIRCLE_SHA1 }}
+          paths:
+            - "~/docker-cache"
+      - deploy:
+          command: |
+            echo "${DOCKER_PASSWORD}" | docker login -u "${DOCKER_USERNAME}" --password-stdin
+            docker push cgswong/${APP_NAME}:${APP_VERSION}
```

With no filters set or control flow set, this deploy job will run on ever pushed commit of every branch.

As you are running the entire process from a single job, you might be best served with a bash evaluation:

```suggestion
            if [  "$CIRCLE_BRANCH" = "master" ]; then 
              docker push cgswong/${APP_NAME}:${APP_VERSION};
            else
              echo "Not on master branch, will not deploy to Docker Hub;
            fi
```

If you were to split the one job into multiple jobs (i.e. `build and test` job -> `deploy` job), you could add circle CI filters that will only run the `deploy` job for specific branchs:
```
workflows:
  version: 2
  build-test-deploy:
    jobs:
      - build-test
      - deploy
          requires:
              - build-test
          filters:
              branches:
                  only: master
```